Support Dynamic Primary Pod NIC Name

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Add a mechanism to dynamically determine the primary pod interface for KuveVirt virtual machines.
The goal is to move away from the hard coded eth0 interface and allow flexible, environment-driven interface
selection.

A new field is added to `v1.VirtualMachineInstanceNetworkInterface` in order to report the matching pod interface name.

VMI controller's network `UpdateStatus` logic is adjusted to report the primary pod interface name inferred from the pod's `k8s.v1.cni.cncf.io/network-status` annotation, with a fallback to the default `eth0`.

virt-handler's network status reporting logic is adjusted to preserve the primary pod interface name.
virt-handler and virt-launcher network setup logic is adjusted to consume a custom primary pod interface name.

Additionally, a small change is made to the network e2e tests to account for addition of the new field.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->
Design proposal: https://github.com/kubevirt/community/pull/324

### Special notes for your reviewer
~~Depends on PR #13038, please skip the first two commits.~~
First three commits are small refactoring.

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [X] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [X] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support Dynamic Primary Pod NIC Name
```

orelmisan commented:
/test pull-kubevirt-e2e-kind-sriov

oshoval commented:
nice, looks good
```
  interfaces:
  - infoSource: domain
    ipAddress: 10.100.0.6
    ipAddresses:
    - 10.100.0.6
    mac: 52:54:00:db:37:70
    name: passtnet
    podInterfaceName: ovn-udn1
    queueCount: 1
```
Please see the gists below how to test it once needed
https://gist.github.com/oshoval/f664e6c650f2167e9248c6cf0cbc30e7
https://gist.github.com/RamLavi/91a25858a56f87e47039ceef99df662b (use first example)

maiqueb commented on pkg/network/multus/status_test.go:
```diff
@@ -33,72 +32,133 @@ import (
+		},
+			Entry("when network status is nil", nil),
+			Entry("when network status is empty", []networkv1.NetworkStatus{}),
+			Entry("when network status does not report interface name",
```
should we add an entry that is "legit" (i.e. features the interface name) but is *not* the default net ?

orelmisan replied:
Could you please give an example of such a scenario? (preferably the JSON value)

orelmisan replied:
Done.

maiqueb commented on pkg/network/multus/status_test.go:
```diff
@@ -33,72 +32,133 @@ import (
+			Entry("When the primary pod interface name is reported and its the default value",
+				[]networkv1.NetworkStatus{
+					{Name: "k8s-pod-network", Interface: defaultPrimaryPodIfaceName, Default: true},
+					{Name: "some-net", Interface: "pod123456", Default: false},
```
nit: no need to define the `Default` attribute, right ? 

It's opinionated, your call.

orelmisan replied:
Could you please elaborate why not?

nirdothan replied:
@maiqueb isn't that the idea per [proposed design] ?(https://github.com/kubevirt/community/blob/2bebe5193a88ebdafcc3c2f1b3432d9c3ac726d0/design-proposals/dynamic-primary-pod-nic-name.md?plain=1#L192)

maiqueb replied:
> I guess you mean because it is the default.

yes, because booleans default to false.

> I prefer to explicitly state the Default field's value - as it directly affects the logic.

all good, I honestly don't care.

maiqueb commented on pkg/network/multus/status_test.go:
```diff
@@ -33,72 +32,133 @@ import (
+			Entry("When the primary pod interface name is reported and its the custom value",
+				[]networkv1.NetworkStatus{
+					{Name: "k8s-pod-network", Interface: customPrimaryPodIfaceName, Default: true},
+					{Name: "some-net", Interface: "net1", Default: false},
```
ditto

oshoval commented:
> - infoSource: domain, guest-agent

@orelmisan is it fine that multus is not there?
it is expected to be just those 2 ?

orelmisan commented:
> > * infoSource: domain, guest-agent
> 
> @orelmisan is it fine that multus is not there? is it expected to be just those 2 ?

Yes.

orelmisan commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-network
/test pull-kubevirt-e2e-kind-sriov

orelmisan commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-network
/test pull-kubevirt-e2e-kind-sriov

oshoval commented:
you can temporary bump kubevirtci in this PR to have new multus to fix those errors maybe

qinqon commented:
Have also test it with Edy's PR and it looks ok and it's selecting the proper interface.
/lgtm

oshoval commented:
maybe we should test OVN k8s CI with both this PR + Edy PR to see nothing broke ?
or it can just wait that those PRs are merged

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -119,13 +124,37 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 
+	interfacesStatus = append(interfacesStatus, reportMissingNetworks(vmi.Spec.Networks, interfacesStatus)...)
+
+	reportPodIfaceNames(interfacesStatus, podIfacesNamesByNetworkNames)
```
I think the pattern followed in this function is to return the data and not to show it is mutated.
Even if you do mutate it, I think you should return the slice back.

orelmisan replied:
This code was removed.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -119,13 +124,37 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
+	var missing []v1.VirtualMachineInstanceNetworkInterface
+
+	for _, network := range networks {
+		if is := netvmispec.LookupInterfaceStatusByName(interfacesStatus, network.Name); is != nil {
```
`is` is not needed.

orelmisan replied:
This code was removed.

EdDev commented on pkg/network/setup/netstat_test.go:
```diff
@@ -61,28 +62,34 @@ var _ = Describe("netstat", func() {
 			),
 		).To(Succeed())
 
+		setup.Vmi.Status.Interfaces = []v1.VirtualMachineInstanceNetworkInterface{
```
I do not understand this addition. Why is it needed?

In general, I do not understand why you had to add this podIface thing into the existing tests.
You could have tested it separately, right?
As there is no `network-status` I would expect everything to just continue to pass.

For the cases where the information is available, you can add the relevant tests per the conditions the production code presents.

orelmisan replied:
This code was removed.
Unit tests for this file are currently missing.

EdDev commented on pkg/network/vmispec/interface.go:
```diff
@@ -188,3 +188,22 @@ func HasBindingPluginDeviceInfo(iface v1.Interface, bindingPlugins map[string]v1
+	podIfacesByNetName := map[string]string{}
+
+	for _, network := range networks {
+		is := LookupInterfaceStatusByName(interfaceStatus, network.Name)
```
You can use `i` or `iface` or `ifaceStatus`. `is` is hard for me to process.

orelmisan replied:
Done.

EdDev commented on pkg/network/vmispec/interface.go:
```diff
@@ -188,3 +188,22 @@ func HasBindingPluginDeviceInfo(iface v1.Interface, bindingPlugins map[string]v1
+		}
+
+		if is.PodInterfaceName == "" {
+			return nil, fmt.Errorf("pod interface name is missing for network %q", is.Name)
```
This does not seem right.
You could ignore it or do nothing... I do not think you can declare it is an error.
But if you think you should, please elaborate why.

orelmisan replied:
This code was removed.

EdDev commented on pkg/network/setup/netconf.go:
```diff
@@ -77,6 +77,11 @@ func NewNetConfWithCustomFactoryAndConfigState(nsFactory nsFactory, cacheCreator
 
 // Setup applies (privilege) network related changes for an existing virt-launcher pod.
 func (c *NetConf) Setup(vmi *v1.VirtualMachineInstance, networks []v1.Network, launcherPid int, preSetup func() error) error {
+	podIfaceNamesByNetworkNames, err := vmispec.IndexPodIfacesNamesByNetworkName(vmi.Spec.Networks, vmi.Status.Interfaces)
```
Please place it close to its usage.

EdDev commented on pkg/network/setup/netconf_test.go:
```diff
@@ -121,9 +128,35 @@ var _ = Describe("netconf", func() {
 		Expect(netConf.Setup(vmi, vmi.Spec.Networks, launcherPid, netPreSetupDummyNoop)).NotTo(Succeed())
 	})
 
+	DescribeTable("fails the setup run when interface status is missing", func(interfaceStatues []v1.VirtualMachineInstanceNetworkInterface) {
```
I do not understand why it should fail on this.
As I see it, it should ignore such network and not run setup on them.

When the information is available, then it should do it.

EdDev replied:
You can resolve this.

We should fail and not ignore because we do not what this to reach the virt-launcher SyncVMI.

EdDev commented on pkg/network/setup/netpod/netpod.go:
```diff
@@ -78,6 +78,8 @@ type NetPod struct {
 
 	log *log.FilteredLogger
+
+	podIfaceNamesByNetworkNames map[string]string
```
nit: I think you can drop the `Name`: `podIfaceNamesByNetwork`

EdDev commented on pkg/network/setup/netpod/netpod.go:
```diff
@@ -129,6 +131,12 @@ func WithLogger(logger *log.FilteredLogger) option {
 	}
 }
 
+func WithPodIfaceNamesByNetworkNames(podIfaceNamesByNetworkNames map[string]string) option {
```
Same

EdDev reviewed (changes requested):
Very nice!

I think the main topic for discussion is about the consumption logic behavior when the new information is missing.
The implementation now prefers to return an error and hopefully solve it in the next reconcile loop.
I would like to explore the option of not handling such an interface/network, skipping it until the next reconcile.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a930e2734f168049267935bdfca2fa1da04cf822..d7cf891c766dc14d8edf983ae87b80e89e81d93e): Rebase on top of #13038.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/d7cf891c766dc14d8edf983ae87b80e89e81d93e..015ca3d3428dab5e0b4e28104569e03d79116244): Rebase on top of https://github.com/kubevirt/kubevirt/pull/13038.

EdDev commented on pkg/network/namescheme/networknamescheme.go:
```diff
@@ -138,6 +138,16 @@ func PodHasOrdinalInterfaceName(podNetworkStatus map[string]networkv1.NetworkSta
 
+// PodHasOrdinalInterfaceName2 checks if the given pod network status has at least one pod interface with ordinal name
+func PodHasOrdinalInterfaceName2(podNetworkStatuses []networkv1.NetworkStatus) bool {
+	for _, networkStatus := range podNetworkStatuses {
```
We do not index for performance, we index because it hurts our thinking to see searches being repeated again and again. Indexing performs the mapping at one step and then the usages can reuse it.

I guess we got used to this pattern so we keep using it.

Copying if the range for loop is not always expensive, it depends on the amount of data. I think there is a linter that raises an error when we copy large structures.
In the range loop it is even cheaper, as the variable is re-used and not re-allocated all the time.

Bottom line, we only use indexing if we need to access the original data and mutate it (either in the loop or when we return a pointer to an entry).

orelmisan commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-network
/test pull-kubevirt-e2e-kind-sriov

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/5a70aafa468312ba2b63b645ad00993a1661367e..535122dcc81017f9276dc2f74e42f8d335f41689): Fixed a NIC hotunplug issue (and a small nit).

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/535122dcc81017f9276dc2f74e42f8d335f41689..70dbd71ddefa0aef3fb2760ef219b1a963721d90): Rebase

orelmisan commented on staging/src/kubevirt.io/api/core/v1/types.go:
Done.

orelmisan commented on pkg/network/setup/netstat.go:
```diff
@@ -119,13 +124,37 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 
 	interfacesStatus = ifacesStatusFromMultus(interfacesStatus, multusStatusNetworksByName, vmiInterfacesSpecByName)
 
+	interfacesStatus = append(interfacesStatus, reportMissingNetworks(vmi.Spec.Networks, interfacesStatus)...)
```
This code was removed.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -90,60 +90,71 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 	}
 
-	multusStatusNetworksByName := netvmispec.IndexInterfaceStatusByName(
+	oldIfaceStatusesByName := netvmispec.IndexInterfaceStatusByName(
```
nit: `old` is so negative... probably because I am old :)

Maybe prefix with `orig`or `prev`.

orelmisan replied:
This change was reverted.
Nevertheless, `prev` is used instead of `old` in applicable places.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -90,60 +90,71 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 
-	interfacesStatus := ifacesStatusFromDomainInterfaces(domain.Spec.Devices.Interfaces)
-	interfacesStatus = append(interfacesStatus,
+	updatedIfaceStatuses := ifacesStatusFromDomainInterfaces(domain.Spec.Devices.Interfaces)
```
nit: `new`,`fresh` and `current` are also nice alternatives for the prefix.
But I do not think you should have changed it. It makes the review harder and it adds no real value IMO.

orelmisan replied:
This change was reverted.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -90,60 +90,71 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
-func ifacesStatusFromMultus(
-	interfacesStatus []v1.VirtualMachineInstanceNetworkInterface,
-	multusStatusNetworksByName map[string]v1.VirtualMachineInstanceNetworkInterface,
+func restoreVirtCtrlOwnedInfo(
```
I think this is worth explaining what and why it does it. It is not trivial for a reader and I am sure we our-self will not remember much in a year from now.

orelmisan replied:
This change was reverted.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -90,60 +90,71 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
+	for oldIfaceName, oldIfaceStatus := range oldIfaceStatusesByName {
+		updatedIfaceStatus := netvmispec.LookupInterfaceStatusByName(updatedIfaceStatuses, oldIfaceName)
+
+		ifaceSpec, oldIfaceExistsInSpec := vmIfacesSpecByName[oldIfaceName]
```
How about `oldIfaceExistsInSpec` ==> `ifaceSpecExists`. ?

The `if ifaceSpecExists && ifaceSpec.State != v1.InterfaceStateAbsent` reads better and we do not get confused.

orelmisan replied:
This change was reverted.

EdDev commented on pkg/network/setup/netconf.go:
```diff
@@ -77,6 +77,11 @@ func NewNetConfWithCustomFactoryAndConfigState(nsFactory nsFactory, cacheCreator
 func (c *NetConf) Setup(vmi *v1.VirtualMachineInstance, networks []v1.Network, launcherPid int, preSetup func() error) error {
+	podIfaceNamesByNetworkName, err := vmispec.IndexPodIfaceNamesByNetworkName(vmi.Spec.Networks, vmi.Status.Interfaces)
+	if err != nil {
+		return fmt.Errorf("missing pod interface naming info: %v", err)
```
nit: Please add the `vmi.Status.Interfaces` list to the error.

orelmisan replied:
This change was reverted.

EdDev commented on pkg/network/vmispec/interface_test.go:
```diff
@@ -270,6 +270,105 @@ var _ = Describe("VMI network spec", func() {
+
+		It("Should return an empty map when there are no networks an no interface statuses", func() {
+			res, err := netvmispec.IndexPodIfaceNamesByNetworkName([]v1.Network{}, []v1.VirtualMachineInstanceNetworkInterface{})
+			Expect(err).NotTo(HaveOccurred())
```
nit: This is not needed if you do this:

```go
Expect(netvmispec.IndexPodIfaceNamesByNetworkName([]v1.Network{}, []v1.VirtualMachineInstanceNetworkInterface{})).To(BeEmpty())
```

The error check is implicit.

I see this is possible to make use of this in other assertions below.
Having direct assertions with less variables is easier to read IMO, but that may be opinionated so it is not a must.

EdDev commented on pkg/network/vmispec/interface.go:
```diff
@@ -193,17 +193,26 @@ func IndexPodIfaceNamesByNetworkName(networks []v1.Network, interfaceStatuses []
 	return podIfacesByNetName, nil
 }
+
+func LookupPodIfaceNameFromStatus(networkName string, interfaceStatuses []v1.VirtualMachineInstanceNetworkInterface) (string, error) {
```
Any reason this is exposed?

orelmisan replied:
It is used by the virt-launcher in the N-1 commit.

orelmisan replied:
This change was reverted.

EdDev reviewed (commented):
Very nice, thank you!

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/70dbd71ddefa0aef3fb2760ef219b1a963721d90..611c7dd4be33831783e6477fea51f1426952a87d): Try to stabilize the status reporting logic.

orelmisan commented:
/test all

nirdothan reviewed (changes requested):
Please make sure that the e2e test passes. See comment

nirdothan commented on tests/network/vmi_infosource.go:
```diff
@@ -104,23 +105,26 @@ var _ = SIGDescribe("Infosource", func() {
+					InfoSource:       netvmispec.InfoSourceDomain,
+					MAC:              primaryInterfaceMac,
+					Name:             primaryNetwork,
+					PodInterfaceName: namescheme.PrimaryPodInterfaceName,
```
The test seems to fail becuse `podInterfaceName` is not as expected 

```
the missing elements were
      <[]v1.VirtualMachineInstanceNetworkInterface | len:3, cap:3>: [
          {
              IP: "",
              MAC: "02:00:05:05:05:05",
              Name: "default",
              IPs: nil,
              PodInterfaceName: "eth0",
              InterfaceName: "",
              InfoSource: "domain",
              QueueCount: 1,
          },
          {
              IP: "",
              MAC: "02:00:a0:a0:a0:a0",
              Name: "bridge1-unchanged",
              IPs: nil,
              PodInterfaceName: "pod678b96116fb",
              InterfaceName: "eth1",
              InfoSource: "domain, guest-agent, multus-status",
              QueueCount: 1,
          },
          {
              IP: "",
              MAC: "02:00:a1:a1:a1:a1",
              Name: "bridge2-setns",
              IPs: nil,
              PodInterfaceName: "pod5a30c9d2cef",
              InterfaceName: "",
              InfoSource: "domain, multus-status",
              QueueCount: 1,
          },
      ]
 ```

  ```
  the extra elements were
      <[]v1.VirtualMachineInstanceNetworkInterface | len:3, cap:3>: [
          {
              IP: "",
              MAC: "02:00:05:05:05:05",
              Name: "default",
              IPs: nil,
              PodInterfaceName: "",
              InterfaceName: "",
              InfoSource: "domain",
              QueueCount: 1,
          },
          {
              IP: "",
              MAC: "02:00:a0:a0:a0:a0",
              Name: "bridge1-unchanged",
              IPs: nil,
              PodInterfaceName: "",
              InterfaceName: "eth1",
              InfoSource: "domain, guest-agent, multus-status",
              QueueCount: 1,
          },
          {
              IP: "",
              MAC: "02:00:a1:a1:a1:a1",
              Name: "bridge2-setns",
              IPs: nil,
              PodInterfaceName: "",
              InterfaceName: "",
              InfoSource: "domain, multus-status",
              QueueCount: 1,
          },
      ]

```

oshoval replied:
you need new kubevirtci for new multus (not bumped yet as part of kubevirt)
you can cherry-pick it manually meanwhile in this PR

EdDev replied:
> you need new kubevirtci for new multus (not bumped yet as part of kubevirt) you can cherry-pick it manually meanwhile in this PR

@oshoval , do you see the bug that the new multus version is fixing manifested here?

oshoval replied:
in order to find the default interface, you need to have the default field correct, and this was fixed only on multus 4.1.1 as far as i understand

oshoval replied:
just cherry-pick this one to try
https://github.com/kubevirt/kubevirt/pull/13041
note that it will break other tests of network 

so you can also create a provider that removes calico bump meanwhile if you want to have it clean (temporary of course)

EdDev replied:
> in order to find the default interface, you need to have the default field correct, and this was fixed only on multus 4.1.1 as far as i understand

The theory is not always in sync with reality. I do not see a problem with the mentioned multus issue, it does not seem to be manifested here.
I guess the diff shown is from this run failure: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/13018/pull-kubevirt-e2e-k8s-1.31-sig-network/1844356434607738880
You can double check to confirm it, maybe I did miss something.

EdDev replied:
> this test is failing, either the test is wrong or we have a problem with the population of `podInterfaceName`

I am guessing that the PR is in draft because there are still issues to solve.
Until it is un-draft , we can just try to find the problems and suggest solutions.

orelmisan replied:
Hi @oshoval, the failure is not related to the Multus version being used.
It was a failure with the new logic.

EdDev replied:
@orelmisan , please resolve this thread.

nirdothan reviewed (changes requested):
this test is failing, either the test is wrong or we have a problem with the population of `podInterfaceName`

orelmisan commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-network
/test pull-kubevirt-e2e-kind-sriov

orelmisan commented on pkg/network/vmicontroller/status.go:
```diff
@@ -32,28 +34,87 @@ import (
+
+	networkStatuses := multus.NetworkStatusesFromPod(pod)
+
+	if primaryInterfaceStatus := calculatePrimaryInterfaceStatus(vmi, networkStatuses); primaryInterfaceStatus != nil {
```
Done.

orelmisan commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-network
/test pull-kubevirt-e2e-kind-sriov

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/aed83ddd0352994cfe6a2d22653442cbf2dfdc85..08f9a976e539c2cca4bf57c62643ae49cc0b78ea): Decrease the scope of the PR to only handle dynamic primary interface name.
The propagation of secondary interfaces' naming scheme was dropped and will be handled in a follow-up.

orelmisan commented on pkg/network/setup/netstat_test.go:
```diff
@@ -110,8 +110,8 @@ var _ = Describe("netstat", func() {
 			Expect(setup.NetStat.UpdateStatus(setup.Vmi, setup.Domain)).To(Succeed())
 
 			Expect(setup.Vmi.Status.Interfaces).To(Equal([]v1.VirtualMachineInstanceNetworkInterface{
-				newVMIStatusIface(primaryNetworkName, []string{primaryPodIPv4, primaryPodIPv6}, "", "", netvmispec.InfoSourceDomain, netsetup.DefaultInterfaceQueueCount),
```
Missing unit tests for the new logic.

orelmisan replied:
Done.

orelmisan commented on pkg/network/setup/netpod/netpod.go:
```diff
@@ -64,12 +64,13 @@ type NSExecutor interface {
-	queuesCap     int
+	vmiSpecIfaces    []v1.Interface
+	vmiSpecNets      []v1.Network
+	vmiIfaceStatuses []v1.VirtualMachineInstanceNetworkInterface
```
Missing unit tests for the new logic.

orelmisan replied:
Done.

orelmisan commented on pkg/virt-handler/non-root.go:
```diff
@@ -117,6 +118,13 @@ func getTapDevices(vmi *v1.VirtualMachineInstance, networkBindings map[string]v1
 	}
 
 	networkNameScheme := namescheme.CreateHashedNetworkNameScheme(vmi.Spec.Networks)
+	if primaryNetwork := vmispec.LookUpDefaultNetwork(vmi.Spec.Networks); primaryNetwork != nil {
```
Missing unit tests for the new logic.

orelmisan commented on pkg/network/link/discovery_test.go:
```diff
@@ -59,13 +59,13 @@ var _ = Describe("DiscoverByNetwork", func() {
 
 	It("should fail when the given no networks", func() {
-		_, err := DiscoverByNetwork(mockNetworkHandler, nil, v1.Network{Name: "ensp1f2"})
+		_, err := DiscoverByNetwork(mockNetworkHandler, nil, v1.Network{Name: "ensp1f2"}, []v1.VirtualMachineInstanceNetworkInterface{})
```
Missing unit tests for the new logic.

orelmisan replied:
Done.

orelmisan commented on pkg/network/namescheme/networknamescheme_test.go:
```diff
@@ -232,7 +232,7 @@ var _ = Describe("Network Name Scheme", func() {
 	Context("HashedPodInterfaceName", func() {
 		DescribeTable("should return the given network name's hashed pod interface name",
 			func(network virtv1.Network, expectedPodIfaceName string) {
-				Expect(namescheme.HashedPodInterfaceName(network)).To(Equal(expectedPodIfaceName))
```
Missing unit tests for the new logic.

Acedus commented on pkg/network/namescheme/networknamescheme.go:
```diff
@@ -138,6 +138,16 @@ func PodHasOrdinalInterfaceName(podNetworkStatus map[string]networkv1.NetworkSta
 	return false
 }
 
+// PodHasOrdinalInterfaceName2 checks if the given pod network status has at least one pod interface with ordinal name
```
What is the end goal of `PodHasOrdinalInterfaceName2`? Is it supposed to supplant the original `PodHasOrdinalInterfaceName2`?

If it is a replacement and you're aiming to supplant the latter, consider adding a deprecation notice so that people prefer using `PodHasOrdinalInterfaceName2`. Personally I think the additional effort of establishing a distinction between the two is worth it (e.g., `PodHasOrdinalIfaceNameInMap`, `PodHasOrdinalIfaceNameInSlice`)

orelmisan replied:
> What is the end goal of `PodHasOrdinalInterfaceName2`? Is it supposed to supplant the original `PodHasOrdinalInterfaceName`?

A follow-up PR to replace the original `PodHasOrdinalInterfaceName` with this one is in the works.

> If it is a replacement and you're aiming to supplant the latter, consider adding a deprecation notice so that people prefer using `PodHasOrdinalInterfaceName2`. Personally I think the additional effort of establishing a distinction between the two is worth it (e.g., `PodHasOrdinalIfaceNameInMap`, `PodHasOrdinalIfaceNameInSlice`) regardless.

Since the number contribution to sig-network related code is small, and the follow-up is in the works, I don't think its worth the effort to mark the original function as deprecated.

Acedus commented on pkg/network/vmicontroller/status.go:
```diff
@@ -34,15 +34,49 @@ import (
-	interfaceStatuses, err := calculateSecondaryIfaceStatuses(vmi, multus.NetworkStatusesFromPod(pod))
+	var interfaceStatuses []v1.VirtualMachineInstanceNetworkInterface
+
+	networkStatuses := multus.NetworkStatusesFromPod(pod)
```
To me this raises a question to me of whether assertion should continue if the length of `networkStatuses` is `<= 1`.

orelmisan replied:
`multus.NetworkStatusesFromPod(pod)` can return an empty slice to a slice with N elements.
The functions which calculate the interfaces statuses take it into account.

Acedus commented on pkg/network/vmicontroller/status.go:
```diff
@@ -34,15 +34,49 @@ import (
+	var interfaceStatuses []v1.VirtualMachineInstanceNetworkInterface
+
+	networkStatuses := multus.NetworkStatusesFromPod(pod)
+
```
nit: consider changing the computation order of `calculatePrimaryIfaceStatus` and `calculateSecondaryIfaceStatus` as the worst case scenario of the latter yields an error and voids the result of the former.

orelmisan replied:
There is an importance to the ordering here:
We want the primary interface to be first in the interface statuses slice, since its IP value is considered to be the VMI's IP address:
https://github.com/kubevirt/kubevirt/blob/ca0227a4c9aee3e6ad16b449ee57bea5325ba998/pkg/virt-operator/resource/generate/components/crds.go#L170

There is similar logic in the virt-handler's network status reporting logic.

EdDev replied:
Our ability to read and compute the flow of changes is also more important than the optimization of a few cycles that may be saved if an error will occur.
I would advice never to prefer that over readability.

BTW, without entering a long debate here, both the compiler or CPU may process this in a different order, use cashes and many other techniques it may find fit.
The code better express the intent and less be worried about optimizations like these.
Said that, I am many times bothered if a search is done N times instead of once, so there is never a total truth to this game.

Acedus replied:
Thanks for the clarification, I concur that premature optimization should be avoided in general, especially in favor of readability.

ormergi commented on pkg/network/multus/status_test.go:
```diff
@@ -33,72 +33,111 @@ import (
+			result := multus.NonDefaultNetworkStatusIndexedByIfaceName(newStubPod(annotations))
+			Expect(result).To(BeEmpty())
+		},
+			Entry("when network-status is missing", map[string]string{}),
```
Does it make sense to test a scenario where the annotation value is invalid here as well?

orelmisan replied:
When the annotation value is invalid, `multus.NetworkStatusesFromPod()` will return an empty slice.
This scenario is covered.

EdDev commented on pkg/network/setup/netstat.go:
```diff
@@ -112,13 +112,29 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
-
 	interfacesStatus = ifacesStatusFromMultus(interfacesStatus, multusStatusNetworksByName, vmiInterfacesSpecByName)
 
+	if vmiSpecPrimaryNet := netvmispec.LookUpDefaultNetwork(vmi.Spec.Networks); vmiSpecPrimaryNet != nil {
```
IIUC the logic below, unlike before, given that the primary network is defined, you always add it to the status, even if it was not detected before (by the domain) or by the vmi controller.

First, I do not understand how this can even occur, but even if it may occur, I do not think we should "lie" about it. We should report only what we detect and preserve specific info (e.g. `podInterfaceName`) from the vmi controller.

I would have expected something like this:
```go
primaryInterfaceStatus, interfacesStatus := netvmispec.PopInterfaceByNetwork(interfacesStatus, netvmispec.LookupPodNetwork(vmi.Spec.Networks))
if primaryInterfaceStatus == nil && prevPrimaryIfaceStatus != nil && vmiSpecPrimaryNet != nil {
        primaryInterfaceStatus = new(v1.VirtualMachineInstanceNetworkInterface)
}
if primaryInterfaceStatus != nil { 
    primaryInterfaceStatus.PodInterfaceName = prevPrimaryIfaceStatus.PodInterfaceName
    interfacesStatus = append([]v1.VirtualMachineInstanceNetworkInterface{*primaryInterfaceStatus}, interfacesStatus...)
}
```

orelmisan replied:
The logic was fixed.

ormergi reviewed (changes requested):
Review up to commit "vmi status, interfaces: Report primary interface"

EdDev commented on pkg/network/link/discovery_test.go:
```diff
@@ -59,13 +59,13 @@ var _ = Describe("DiscoverByNetwork", func() {
 
 	It("should fail when the given no networks", func() {
-		_, err := DiscoverByNetwork(mockNetworkHandler, nil, v1.Network{Name: "ensp1f2"})
+		_, err := DiscoverByNetwork(mockNetworkHandler, nil, v1.Network{Name: "ensp1f2"}, []v1.VirtualMachineInstanceNetworkInterface{})
```
Why you are not passing in a `nil`?

orelmisan replied:
Done.

orelmisan commented on pkg/network/namescheme/networknamescheme.go:
```diff
@@ -138,6 +138,16 @@ func PodHasOrdinalInterfaceName(podNetworkStatus map[string]networkv1.NetworkSta
 }
 
+// PodHasOrdinalInterfaceName2 checks if the given pod network status has at least one pod interface with ordinal name
+func PodHasOrdinalInterfaceName2(networkStatuses []networkv1.NetworkStatus) bool {
```
This issue was raised twice, please see:
- https://github.com/kubevirt/kubevirt/pull/13018#discussion_r1793291246
- https://github.com/kubevirt/kubevirt/pull/13018#discussion_r1799014788

It will be fixed in a follow-up.

orelmisan commented on pkg/network/vmicontroller/status.go:
```diff
@@ -32,28 +34,47 @@ import (
 			})
 		case exists && vmiIfaceStatus != nil:
-			vmiIfaceStatus.InfoSource = vmispec.AddInfoSource(vmiIfaceStatus.InfoSource, vmispec.InfoSourceMultusStatus)
+			updatedIfaceStatus := *vmiIfaceStatus
```
It is needed in order to not modify the original values.
Please see https://github.com/kubevirt/kubevirt/pull/13018#discussion_r1793378480.

ormergi replied:
Please resolve this thread

orelmisan commented on pkg/network/vmicontroller/status.go:
```diff
@@ -34,15 +34,49 @@ import (
 	return nil
 }
 
+func calculatePrimaryIfaceStatus(
```
It is tested as part of `UpdateStatus` in `pkg/network/vmicontroller/status_test.go`.

EdDev commented on pkg/network/multus/status_test.go:
```diff
@@ -33,72 +32,117 @@ import (
+				},
+			}
+
+			Expect(multus.NetworkStatusesFromPod(newStubPod(annotations))).To(Equal(expectedResult))
```
> I don't insist on this but I generally think that it is always better practice to use [ConsistOf](https://onsi.github.io/gomega/#consistofelement-interface) when comparing multi item lists. The logic here may change in the future and order may not remain guaranteed,

A list/slice has order, so it is unclear to me why a unit test cannot assure it is maintained and preserved.
I agree it is not always a fit, but it depends on the specification/API of the expectation.

 > furthermore Gomega's message may is better tailored for slices in case of errors
 
For that you could suggest https://onsi.github.io/gomega/#haveexactelementselement-interface.
On which there will be probably no dispute.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/24992bfa7d6f989af14bd75d5e0c6b337ae52953..9c495e86a9c871c042a6e8b1bbb230ac10cf1766): Answered code review comments.
Added unit tests.

[Change2](https://github.com/kubevirt/kubevirt/compare/9c495e86a9c871c042a6e8b1bbb230ac10cf1766..eaefe577cf00e3d94feb57aa2ce3a87b4ba44df2): Fixed signoff.

orelmisan commented on pkg/network/multus/status_test.go:
```diff
@@ -33,67 +33,45 @@ import (
-			Expect(result).To(Equal(expectedResult))
-		})
-
-		It("Should return a map of pod interface name to network status when interface name is ordinal", func() {
```
As the commit message says: The interface name has no meaning to the test.
It could be any name.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/eaefe577cf00e3d94feb57aa2ce3a87b4ba44df2..6286ccdf9d71c62001109579e5567b6a5d81eced): Answered @EdDev's comment.

orelmisan commented:
[Change1](https://github.com/kubevirt/kubevirt/compare/6286ccdf9d71c62001109579e5567b6a5d81eced..d4585412805b8df44862913036319a1e9eee411d), [Change2](https://github.com/kubevirt/kubevirt/compare/d4585412805b8df44862913036319a1e9eee411d..6b59cd342c79d646f5cab6237b0d4a709f55160f): Rename `restoreVirtCtrlOwnedInfo` to `restorePodIfaceNameForPrimaryIface` as there is additional info we restore from virt-controller (Multus info source).

EdDev reviewed (commented):
Thank you!

/approve

orelmisan commented on pkg/virt-handler/non-root.go:
```diff
@@ -117,6 +118,13 @@ func getTapDevices(vmi *v1.VirtualMachineInstance, networkBindings map[string]v1
 	}
 
 	networkNameScheme := namescheme.CreateHashedNetworkNameScheme(vmi.Spec.Networks)
+	if primaryNetwork := vmispec.LookUpDefaultNetwork(vmi.Spec.Networks); primaryNetwork != nil {
```
Done.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/6b59cd342c79d646f5cab6237b0d4a709f55160f..db9446376eb8e42a0308fab654d2c5e0c6501ec6): Answered @xpivarc comments.

ormergi commented on pkg/network/vmicontroller/status_test.go:
```diff
@@ -45,6 +46,29 @@ var _ = Describe("Status Update", func() {
 
 		multusNetworkStatusWithPrimaryNet = `[{"name":"k8s-pod-network","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}}]`
 
+		multusNetworkStatusWithPrimaryNetAndIfaceName = `[` +
+			`{"name":"k8s-pod-network","interface":"eth0","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}}` +
+			`]`
+
+		multusNetworkStatusWithAlternativePrimaryNet = `[` +
+			`{"name":"alternativeNAD", "ips":["10.244.196.146", "fd10:244::c491"], "default":true, "dns":{}}` +
+			`]`
+
+		multusNetworkStatusWithAlternativePrimaryNetAndIfaceName = `[` +
+			`{"name":"alternativeNAD", "interface":"eth0", "ips":["10.244.196.146", "fd10:244::c491"], "default":true,"dns":{}}` +
+			`]`
+
+		multusNetworkStatusWithCustomPrimaryNet = `[` +
+			`{"name":"k8s-pod-network", "interface":"eth0", "ips":["10.244.196.146", "fd10:244::c491"], "default":false,"dns":{}}, ` +
+			`{"name":"cluster-network", "interface":"custom-iface", "ips":["10.128.0.4"], "mac":"0a:58:0a:80:00:04", "default":true, "dns":{}}` +
+			`]`
+
+		multusNetworkStatusWithCustomPrimaryAndSecondaryNets = `[` +
+			`{"name":"k8s-pod-network", "interface":"eth0", "ips":["10.244.196.146", "fd10:244::c491"], "default":false,"dns":{}}, ` +
+			`{"name":"cluster-network", "interface":"custom-iface", "ips":["10.128.0.4"], "mac":"0a:58:0a:80:00:04", "default":true, "dns":{}}, ` +
+			`{"name":"meganet", "interface":"pod7e0055a6880", "mac":"8a:37:d9:e7:0f:18", "dns":{}}` +
+			`]`
+
```
nit: looking at these tests again, I think it would make it easier to review and follow if these const were inlined.

ormergi reviewed (commented):
Overall looks good, functionality wise.

I have problem with some cosmetic stuff such as components naming, documentation, and some tests.
Few chacnges could have simplified but functionallity is fine so I tried ignore them.
Due to time constrains towards  the upcoming FF I am not insisting on these changes this time.
Such comments were marked as 'nit' although there are some comments we do not consider as nits.

orelmisan commented on pkg/network/vmicontroller/status_test.go:
```diff
@@ -120,45 +309,105 @@ var _ = Describe("Status Update", func() {
 	)
 
-	It("Should keep the Multus info source when VMI.status has an interface and it is reported by Multus network-status", func() {
+	DescribeTable("VMI with primary and secondary networks", func(podAnnotations map[string]string, expectedPrimaryInterfaceName string) {
```
Done.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/db9446376eb8e42a0308fab654d2c5e0c6501ec6..7ba63d5594c11642e1c899d9ff59fcc8829e544c): Addressed @ormergi's [comment](https://github.com/kubevirt/kubevirt/pull/13018#discussion_r1801099517).
Removed unused consts.

orelmisan commented on pkg/network/setup/netstat.go:
```diff
@@ -126,6 +129,32 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 	return nil
 }
 
+func restorePodIfaceNameForPrimaryIface(
```
The VMI controller owns two pieces of information with regards to the VMI interfaces status:
1. The `multus-status` info source.
2. The pod's primary interface name.

When virt-handler calculates the VMI interfaces status - it calculates a new slice.
There are two steps it takes to "restore" the VMI controller-owned info:
1. `ifacesStatusFromMultus`.
2. `restorePodIfaceNameForPrimaryIface`.

orelmisan commented on pkg/virt-handler/non-root.go:
```diff
@@ -117,6 +117,8 @@ func getTapDevices(vmi *v1.VirtualMachineInstance, networkBindings map[string]v1
 	}
 
 	networkNameScheme := namescheme.CreateHashedNetworkNameScheme(vmi.Spec.Networks)
+	networkNameScheme = namescheme.UpdatePrimaryPodIfaceNameFromVMIStatus(networkNameScheme, vmi.Spec.Networks, vmi.Status.Interfaces)
```
`namescheme.UpdatePrimaryPodIfaceNameFromVMIStatus` takes a podIfaceNamesByNetworkName map (a naming scheme), and updates its primary pod interface name in case it exists in the VMI.Status.Interfaces slice.

orelmisan commented on pkg/network/link/discovery.go:
```diff
@@ -36,17 +36,17 @@ import (
-	ifaceName := namescheme.HashedPodInterfaceName(subjectNetwork)
+func networkInterfaceNames(networks []v1.Network, subjectNetwork v1.Network, ifaceStatuses []v1.VirtualMachineInstanceNetworkInterface) ([]string, error) {
+	ifaceName := namescheme.HashedPodInterfaceName(subjectNetwork, ifaceStatuses)
 	ordinalIfaceName := namescheme.OrdinalPodInterfaceName(subjectNetwork.Name, networks)
```
I don't follow.
Which function is used here?

xpivarc reviewed (commented):
/approve
/hold
I see we already have a feature gate follow up. I tend to merge this as there are holidays in place and the follow up seems to target all the requests/concerns.

@EdDev Remove hold if you are happy with this from network side.

qinqon commented:
/retest

EdDev commented:
> /approve /hold I see we already have a feature gate follow up. I tend to merge this as there are holidays in place and the follow up seems to target all the requests/concerns.
> 
> @EdDev Remove hold if you are happy with this from network side.

I think we are good.
The FG [PR](https://github.com/kubevirt/kubevirt/pull/13078) is almost ready and should rebase over this once it is merged.

EdDev commented:
/unhold

EdDev commented:
/retest-required
